File: stage3/02-build-provisioning/files/iot-provision.js
#!/usr/bin/env node

const fs = require("fs");
const path = require("path");
const { mqtt, io, iot } = require("aws-iot-device-sdk-v2");

// Eatabit library directory
const EATABIT_ROOT_DIR = "/usr/local/lib/eatabit";

// Claim cert paths (bootstrap)
const EATABIT_CERT_PATH = `${EATABIT_ROOT_DIR}/cert`;
const CLAIM_CERT = `${EATABIT_CERT_PATH}/71cc32e68839f91c3d1f96f5ad42e27bf3450c735b8eb928ebb9a0bfb9fb7235-certificate.pem.crt`;
const CLAIM_KEY = `${EATABIT_CERT_PATH}/71cc32e68839f91c3d1f96f5ad42e27bf3450c735b8eb928ebb9a0bfb9fb7235-private.pem.key`;
const ROOT_CA = `${EATABIT_CERT_PATH}/AmazonRootCA1.pem`;

const AWS_IOT_ENDPOINT = "a3fw1u2gvi2uac-ats.iot.us-east-2.amazonaws.com";
const TEMPLATE_NAME = "templateProvisioning";
const DEVICE_ID = fs
  .readFileSync("/proc/cpuinfo", "utf8")
  .match(/Serial\s*:\s*(\w+)/)[1]; // Pi serial
const VERSION = fs.readFileSync(`${EATABIT_ROOT_DIR}/version`, "utf8").trim();
const AWS_DEVICE_FILE = `${EATABIT_ROOT_DIR}/config/aws-device.json`;

const LOG_FILE = `${EATABIT_ROOT_DIR}/log/provision.log`;

// Topic shortcuts
const TOPIC_CERT_CREATE = "$aws/certificates/create/json";
const TOPIC_CERT_ACCEPTED = "$aws/certificates/create/json/accepted";
const TOPIC_CERT_REJECTED = "$aws/certificates/create/json/rejected";
const TOPIC_REGISTER = `$aws/provisioning-templates/${TEMPLATE_NAME}/provision/json`;
const TOPIC_REGISTER_ACCEPTED = `$aws/provisioning-templates/${TEMPLATE_NAME}/provision/json/accepted`;
const TOPIC_REGISTER_REJECTED = `$aws/provisioning-templates/${TEMPLATE_NAME}/provision/json/rejected`;

const TARGET_CERT_PATH = `${EATABIT_CERT_PATH}/device.pem`;
const TARGET_KEY_PATH = `${EATABIT_CERT_PATH}/device.key`;

// Ensure log directory and file exist
try {
  const logDir = path.dirname(LOG_FILE);
  if (!fs.existsSync(logDir)) {
    fs.mkdirSync(logDir, { recursive: true });
  }
  if (!fs.existsSync(LOG_FILE)) {
    fs.writeFileSync(LOG_FILE, "", { mode: 0o644 });
  }
} catch (err) {
  console.error(`Failed to initialize log file: ${err.message}`);
}

function log(message, level = "INFO") {
  const timestamp = new Date().toISOString().replace("T", " ").slice(0, 19);
  const line = `[${timestamp}] ${level}: ${message}\n`;
  console.log(line.trim());
  try {
    fs.appendFileSync(LOG_FILE, line);
  } catch (err) {
    console.error("Failed to write to log file:", err.message);
  }
}

log(`Starting provisioning with broker: ${AWS_IOT_ENDPOINT}`);

function decodePayload(payload) {
  try {
    if (Buffer.isBuffer(payload)) {
      return payload.toString("utf8");
    }
    if (payload instanceof ArrayBuffer) {
      return Buffer.from(payload).toString("utf8");
    }
    if (ArrayBuffer.isView(payload)) {
      return Buffer.from(
        payload.buffer,
        payload.byteOffset,
        payload.byteLength
      ).toString("utf8");
    }
    return String(payload);
  } catch {
    return String(payload);
  }
}

async function run() {
  const clientBootstrap = new io.ClientBootstrap();
  let provisioningSucceeded = false;

  const configBuilder =
    iot.AwsIotMqttConnectionConfigBuilder.new_mtls_builder_from_path(
      CLAIM_CERT,
      CLAIM_KEY
    );
  configBuilder.with_certificate_authority_from_path(undefined, ROOT_CA);
  configBuilder.with_endpoint(AWS_IOT_ENDPOINT);
  configBuilder.with_client_id(DEVICE_ID);
  configBuilder.with_clean_session(true);

  const config = configBuilder.build();
  const client = new mqtt.MqttClient(clientBootstrap);
  const connection = client.new_connection(config);

  let ownershipToken = null;

  connection.on("connect", async () => {
    log("Connected to AWS IoT Core");

    await connection.subscribe(TOPIC_CERT_ACCEPTED, mqtt.QoS.AtLeastOnce);
    await connection.subscribe(TOPIC_CERT_REJECTED, mqtt.QoS.AtLeastOnce);
    await connection.subscribe(TOPIC_REGISTER_ACCEPTED, mqtt.QoS.AtLeastOnce);
    await connection.subscribe(TOPIC_REGISTER_REJECTED, mqtt.QoS.AtLeastOnce);

    log("Subscribed to provisioning response topics");

    // Step 1: Request new keys and certificate
    await connection.publish(
      TOPIC_CERT_CREATE,
      JSON.stringify({}),
      mqtt.QoS.AtLeastOnce
    );
    log("Published certificate creation request");
  });

  connection.on("message", async (topic, payload) => {
    const message = decodePayload(payload);

    if (topic === TOPIC_CERT_ACCEPTED) {
      log("Certificate creation accepted");

      try {
        const data = JSON.parse(message);
        const certPem = data.certificatePem;
        const privateKey = data.privateKey;
        ownershipToken = data.certificateOwnershipToken;

        // Save certificate and key
        fs.writeFileSync(TARGET_CERT_PATH, certPem + "\n");
        fs.writeFileSync(TARGET_KEY_PATH, privateKey + "\n");

        log("New certificate and private key saved");

        // Step 2: Register the thing
        const registerPayload = {
          certificateOwnershipToken: ownershipToken,
          parameters: {
            DeviceId: DEVICE_ID,
            Version: VERSION,
          },
        };

        await connection.publish(
          TOPIC_REGISTER,
          JSON.stringify(registerPayload),
          mqtt.QoS.AtLeastOnce
        );
        log("Published device registration request");
      } catch (err) {
        log(`Failed to process certificate response: ${err.message}`, "ERROR");
      }
    } else if (topic === TOPIC_CERT_REJECTED) {
      log(`Certificate creation rejected: ${message}`, "ERROR");
      connection.disconnect();
    } else if (topic === TOPIC_REGISTER_ACCEPTED) {
      log("Device registration successful");

      // Save response
      fs.writeFileSync(AWS_DEVICE_FILE, message);
      log(`Registration response saved to ${AWS_DEVICE_FILE}`);

      // Delete claim certificate and key
      try {
        fs.unlinkSync(CLAIM_CERT);
        fs.unlinkSync(CLAIM_KEY);
        log("Claim certificate and key deleted");
      } catch (err) {
        log(`Warning: Failed to delete claim credentials: ${err.message}`, "WARN");
      }

      log("Provisioning completed successfully");
      provisioningSucceeded = true;
      connection.disconnect();
    } else if (topic === TOPIC_REGISTER_REJECTED) {
      log(`Device registration rejected: ${message}`, "ERROR");
      connection.disconnect();
    }
  });

  connection.on("error", (error) => {
    log(`Connection error: ${error.message}`, "ERROR");
  });

  connection.on("interrupt", () => {
    log("Connection interrupted");
  });

  connection.on("close", () => {
    log("Connection closed");
    process.exit(provisioningSucceeded ? 0 : 1);
  });

  // Timeout to prevent hanging indefinitely
  const PROVISION_TIMEOUT_MS = 60000;
  setTimeout(() => {
    log("Provisioning timed out after 60 seconds", "ERROR");
    process.exit(1);
  }, PROVISION_TIMEOUT_MS);

  await connection.connect();
}

// Run and handle unhandled errors
run().catch((err) => {
  log(`Unexpected error: ${err.message}`, "ERROR");
  process.exit(1);
});

File: tmp/iot-provision.py
#!/usr/bin/env python3

import json
import os
import argparse
from awsiotsdk.mqtt_connection_builder import mtls_from_path
from awsiotsdk import iotshadow
from awscrt import io


# Get the Raspberry Pi serial number as unique client ID
def get_device_serial():
    try:
        with open("/proc/cpuinfo", "r") as f:
            for line in f:
                if line.startswith("Serial"):
                    return line.split(":")[1].strip()
    except:
        return "unknown-device"


# Paths to your provisioning claim (bootstrap) credentials
EATABIT_LIB_DIR = "/usr/local/lib/eatabit"
CLAIM_CERT = f"{EATABIT_LIB_DIR}/cert/71cc32e68839f91c3d1f96f5ad42e27bf3450c735b8eb928ebb9a0bfb9fb7235-certificate.pem.crt"
CLAIM_KEY = f"{EATABIT_LIB_DIR}/cert/71cc32e68839f91c3d1f96f5ad42e27bf3450c735b8eb928ebb9a0bfb9fb7235-private.pem.key"
ROOT_CA = f"{EATABIT_LIB_DIR}/cert/AmazonRootCA1.pem"
ENDPOINT = "a75p0fsmm0h6r-ats.iot.us-east-1.amazonaws.com"
CLIENT_ID = get_device_serial()
TEMPLATE_NAME = "templateProvisioning"

# Directories to save new certs/keys
CERT_DIR = f"{EATABIT_LIB_DIR}/cert"
os.makedirs(CERT_DIR, exist_ok=True)


def on_publish_create_keys_accepted(packet):
    print("CreateKeysAndCertificate accepted")
    payload = json.loads(packet.payload)
    cert = payload["certificatePem"]
    key = payload["keyPair"]["PrivateKey"]
    ownership_token = payload["certificateOwnershipToken"]

    cert_path = os.path.join(CERT_DIR, "device.pem.crt")
    key_path = os.path.join(CERT_DIR, "private.pem.key")

    with open(cert_path, "w") as f:
        f.write(cert)
    with open(key_path, "w") as f:
        f.write(key)

    print(f"Saved new certificate to {cert_path} and key to {key_path}")

    # Now register the thing
    register_payload = {
        "certificateOwnershipToken": ownership_token,
        "parameters": {"DeviceId": CLIENT_ID, "Version": "0.0.1"},
    }
    connection.publish(
        topic=f"$aws/provisioning-templates/{TEMPLATE_NAME}/provision/json",
        payload=json.dumps(register_payload),
        qos=1,
    )


def on_publish_register_thing_accepted(packet):
    print("RegisterThing accepted")
    payload = json.loads(packet.payload)
    print("Provisioning successful!")
    print("Thing Name:", payload.get("thingName"))
    # You can now disconnect and reconnect using the new cert/key for normal operations
    connection.disconnect()


def on_publish_rejected(packet, topic):
    print(f"{topic} rejected")
    payload = json.loads(packet.payload)
    print("Error:", payload)


if __name__ == "__main__":
    # Build MQTT connection using claim cert
    io.init_logging(getattr(io.LogLevel, "Info"), "stderr")
    connection = mtls_from_path(
        endpoint=ENDPOINT,
        cert_filepath=CLAIM_CERT,
        pri_key_filepath=CLAIM_KEY,
        ca_filepath=ROOT_CA,
        client_id=CLIENT_ID,
    )

    print("Connecting with provisioning claim certificate...")
    connect_future = connection.connect()
    connect_future.result()
    print("Connected!")

    # Subscribe to response topics
    create_keys_accepted, _ = connection.subscribe(
        topic="$aws/certificates/create/json/accepted",
        qos=1,
        callback=on_publish_create_keys_accepted,
    )

    create_keys_rejected, _ = connection.subscribe(
        topic="$aws/certificates/create/json/rejected",
        qos=1,
        callback=lambda packet: on_publish_rejected(packet, "CreateKeysAndCertificate"),
    )

    register_accepted, _ = connection.subscribe(
        topic=f"$aws/provisioning-templates/{TEMPLATE_NAME}/provision/json/accepted",
        qos=1,
        callback=on_publish_register_thing_accepted,
    )

    register_rejected, _ = connection.subscribe(
        topic=f"$aws/provisioning-templates/{TEMPLATE_NAME}/provision/json/rejected",
        qos=1,
        callback=lambda packet: on_publish_rejected(packet, "RegisterThing"),
    )

    # Publish to create keys and cert (empty payload for JSON)
    print("Publishing to CreateKeysAndCertificate...")
    connection.publish(topic="$aws/certificates/create/json", payload="{}", qos=1)

    # Keep the script running to receive responses
    print("Waiting for provisioning responses...")
    try:
        while True:
            pass
    except KeyboardInterrupt:
        print("Disconnecting...")
        connection.disconnect().result()
